nuclio_sdk/event.py

```diff
@@ -130,29 +126,102 @@ def from_parsed_data(cls, parsed_data, body, content_type):
+            path=parsed_data[b"path"],
+            size=parsed_data[b"size"],
+            timestamp=datetime.datetime.utcfromtimestamp(parsed_data[b"timestamp"]),
+            url=parsed_data[b"url"],
+            shard_id=parsed_data[b"shard_id"],
+            num_shards=parsed_data[b"num_shards"],
+            _type=parsed_data[b"type"],
+            type_version=parsed_data[b"type_version"],
+            version=parsed_data[b"version"],
+        )
+
+    def __repr__(self):
+        return self.to_json()
+
+
+class EventDeserializer(object):
+    def deserialize(self, event_message) -> Event:
+        raise NotImplementedError
+
+
+class EventDeserializerFactory(object):
     @staticmethod
-    def decode_body(body, content_type):
-        """Decode event body"""
+    def create(deserializer_kind) -> EventDeserializer:
+        if deserializer_kind == "msgpack":
+            return _EventDeserializerMsgPack(raw=False)
+        if deserializer_kind == "msgpack_raw":
+            return _EventDeserializerMsgPack(raw=True)
+        if deserializer_kind == "json":
+            return _EventDeserializerJSON()
+        raise RuntimeError(f"No such deserializer kind {deserializer_kind}")
+
+
+class _EventDeserializerMsgPack(EventDeserializer):
+    def __init__(self, raw=False):
+
+        # return the concrete function that handled raw/decoded event messages
+        # pre-assign to avoid if/else during event processing
+        self._from_msgpack_handler = (
+            self._from_msgpack_raw if raw else self._from_msgpack_decoded
+        )
 
-        if isinstance(body, dict):
-            return body
+    def deserialize(self, event_message: dict):
+        return self._from_msgpack_handler(event_message)
 
-        try:
-            decoded_body = base64.b64decode(body)
-        except:  # noqa E722
-            return body
+    def _from_msgpack_raw(self, parsed_data):
+        body = self._decode_body_raw(parsed_data[b"body"], parsed_data[b"content_type"])
+        return Event.from_parsed_data_bytes(parsed_data, body)
+
+    def _from_msgpack_decoded(self, parsed_data):
+        body = self._decode_body(parsed_data["body"], parsed_data["content_type"])
+        return Event.from_parsed_data(parsed_data, body)
+
+    def _decode_body_raw(self, body, content_type):
+        if content_type == b"application/json":
+            body = self._try_deserialize_json(body)
+
+        return body
 
+    def _decode_body(self, body, content_type):
         if content_type == "application/json":
-            try:
-                return json.loads(decoded_body)
-            except:  # noqa E722
-                pass
+            body = self._try_deserialize_json(body)
 
-        return decoded_body
+        return body
 
-    @staticmethod
-    def decode_msgpack_body(body, content_type):
-        """Decode msgpack event body"""
+    def _try_deserialize_json(self, body):
+        try:
+            return json.loads(body.decode("utf-8"))
```

Comment:
decode in `_decode_body` and pass string to `_try_deserialize_json` (otherwise this function also decodes and the decode does nothing that this doesn't do.
Also, if it's not json it doesn't mean you can skip string decode, right?

Reply:
you actually, found 2 existing bugs
- if nuclio is triggered with a header "application/json" and no body (e.g.: GET), it fails to decode
- when it fails to decode, the error is printed to stderr but it would never be flushed as long as you dont use `\n` (because of how python is executed with `-u` from the processor)

fixing them both now.